libecole/src/reward/dualintegral.cpp

```diff
@@ -0,0 +1,100 @@
+#include "scip/scip.h"
+#include "scip/type_event.h"
+
+#include "ecole/reward/dualintegral.hpp"
+#include "ecole/reward/integral_eventhdlr.hpp"
+#include "ecole/scip/model.hpp"
+#include "ecole/utility/chrono.hpp"
+
+namespace ecole::reward {
+
+namespace {
+
+auto compute_dual_integral(
+	std::vector<SCIP_Real> dual_bounds,
+	std::vector<std::chrono::nanoseconds> times,
+	SCIP_Real dual_bound_reference) {
+	SCIP_Real dual_integral = 0.0;
+	for (size_t i = 0; i < dual_bounds.size() - 1; ++i) {
+		auto const dual_bound_diff = dual_bound_reference - dual_bounds[i];
+		auto const time_diff = std::chrono::duration<double>(times[i + 1] - times[i]).count();
+		dual_integral += dual_bound_diff * time_diff;
+	}
+
+	return dual_integral;
+}
+
+/* Compute the dual bounds adjusted by the initial bound given. */
+auto get_adjusted_dual_bounds(std::vector<SCIP_Real> dual_bounds, SCIP_Real initial_dual_bound) {
+	std::vector<SCIP_Real> adjusted_dual_bounds(dual_bounds.size());
+	for (std::size_t i = 0; i < dual_bounds.size(); ++i) {
+		if (dual_bounds[i] < initial_dual_bound) {
+			adjusted_dual_bounds[i] = initial_dual_bound;
+		} else {
+			adjusted_dual_bounds[i] = dual_bounds[i];
+		}
+	}
```

Comment:
Can you do that on the fly in `compute_dual_integral` so you don't  need to allocate a new vector and iterate over the data twice?

Otherwise, since you are taking parameter by copy taken, you can simply.
```suggestion
auto get_adjusted_dual_bounds(std::vector<SCIP_Real> dual_bounds, SCIP_Real initial_dual_bound) {
	for (auto& b : dual_bounds) {
		if (b < initial_dual_bound) {
			b = initial_dual_bound;
		}
	}
```